sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryQuerySourceDef.java

```diff
@@ -115,19 +127,26 @@ TableReference getTableReference(BigQueryOptions bqOptions, String stepUuid)
         useLegacySql,
         priority,
         location,
+        queryTempDataset,
         kmsKey);
   }
 
   void cleanupTempResource(BigQueryOptions bqOptions, String stepUuid) throws Exception {
+    Optional<String> queryTempDatasetOpt = Optional.ofNullable(queryTempDataset);
     TableReference tableToRemove =
         createTempTableReference(
-            bqOptions.getProject(), createJobIdToken(bqOptions.getJobName(), stepUuid));
+            bqOptions.getProject(),
+            createJobIdToken(bqOptions.getJobName(), stepUuid),
+            queryTempDatasetOpt);
 
     BigQueryServices.DatasetService tableService = bqServices.getDatasetService(bqOptions);
     LOG.info("Deleting temporary table with query results {}", tableToRemove);
     tableService.deleteTable(tableToRemove);
-    LOG.info("Deleting temporary dataset with query results {}", tableToRemove.getDatasetId());
-    tableService.deleteDataset(tableToRemove.getProjectId(), tableToRemove.getDatasetId());
+    if (queryTempDatasetOpt.isPresent()) {
```

Comment:
Probably extract "queryTempDatasetOpt.isPresent()" to an instance variable boolean datasetProvidedByUser (before this point).

Reply:
Ok. Will push a commit with that change, and a bug fix in line 138. The logic in the if is wrong, it should when the optional is *not* present (meaning, that the user did not specify the dataset, and therefore Beam created it).

Reply:
Done.